pkg/filters/pidSet.go

```diff
@@ -22,48 +22,84 @@ import (
 	hubbleFilters "github.com/cilium/tetragon/pkg/oldhubble/filters"
 )
 
-/* pidSet is the set of pids we are interested in receiving events
- * for. This includes the pids listed in filter and any of their
- * children.
- */
-var pidSet map[uint32]bool
+// We could use an LRU here but we really don't want to evict old entries and risk failing
+// a test (where this filter is mostly used). Instead, we simply use a basic Go map and
+// enforce an invariant that we cut off adding new entries if we exceed maxChildCacheSize
+// number of keys. When this happens, we continue to pass the filter on all previously
+// seen PIDs but stop adding new ones, effectively truncating the filter's view of the process tree.
+type ChildCache = map[uint32]struct{}
 
-func filterByPidSet(pids []uint32) hubbleFilters.FilterFunc {
-	return func(ev *v1.Event) bool {
-		process := GetProcess(ev)
-		if process == nil {
-			return false
-		}
-		for _, pid := range pids {
-			if pid == process.Pid.GetValue() {
-				pidSet[pid] = true
-				return true
-			}
-		}
-		parent := GetParent(ev)
-		if parent == nil {
-			return false
-		}
-		if pidSet[parent.Pid.GetValue()] == true {
+const (
+	// Maximum number of keys in a ChildCache before it fills up and we stop adding new
+	// entries.
+	maxChildCacheSize = 8192
+)
+
+func checkPidSetMembership(pid uint32, pidSet []uint32, childCache ChildCache) bool {
+	// Check the original pidSet. The reason for doing this separately is that we never
+	// want to drop the original pidSet from the LRU. Keeping this separately in a slice
+	// is an easy way to achieve this.
+	for _, p := range pidSet {
+		if pid == p {
 			return true
 		}
-		for _, pid := range pids {
-			if pid == parent.Pid.GetValue() {
-				pidSet[pid] = true
-				return true
-			}
-		}
+	}
+	// Fall back to childCache to check children.
+	_, ok := childCache[pid]
```

Comment:
how about:
```
// we reached maximum size, which means that we cannot insert any more pids in thet set, so we always return // true to account for the lost pids.
if len(childCache) == maxChildCacheSize {
              return true
}
```

Reply:
OK, but then for an acceptList, as things stand now, we might loose events, which might lead to flakes. Or am I missing something?

Maybe pass a configuration to the filter so that we make it reliable in both cases?

Reply:
Yeah that's fair. I actually think my old approach using an LRU was maybe slightly better for this, since you'd need 8192 actively running children to start dropping events.

Anyway it's not an issue to make this configurable, so I'll just do it.

Reply:
@kkourt I added a configuration knob for the cache size. FWIW I think 8192 is probably gonna be acceptable in most cases, but now we can at least adjust as needed.